### PR TITLE
feat: no-constant-binary-expression report unnecessary nullish coalescing operator 

### LINKED ISSUE #16408: Rule Change: no-constant-binary-expression should report `<expr1> ?? <non-nullish> ?? <expr2>`
### Rule details

Inform developer of potentially unnecessary nullish coalescing operator

### Related ECMAScript feature

Nullish coalesing operator

### What type of rule is this?

Suggests an alternate way of doing something

### Example code

```js
// fail
window.abc ?? "any non-nullish value" ?? "any non-nullish value" // warn (the last nullish coalesing operator will never be used)
window.abc ?? undefined ?? "any non-nullish value" // warn (the first nullish coalescing operator is skipped)
window.abc ?? null ?? "any non-nullish value" // warn (the first nullish coalescing operator is skipped)
```

### Why should this rule be in the core instead of a plugin?

I don't know.

### Participation

- [ ] I am willing to submit a pull request to implement this rule.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by nissy-dev**
#### Prerequisites checklist

- [ ] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[X] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Fix #16408

I updated the logic of `hasConstantNullishness` for reporting errors in the case of  `<expr1> ?? <non-nullish> ?? <expr2>`

#### Is there anything you'd like reviewers to focus on?

I add some test cases, but there may be edge cases. If you can think of such a case, I would be glad to point it out.

### CODE REVIEW

**Review by snitin315 — COMMENTED** [review]
Thank you for working on this. Can you also update the docs to reflect this case?

**Comment by nissy-dev** [comment]
@snitin315 Thank you for your comments.

> Can you also update the docs to reflect this case?

I added some short-circuits sample cases in docs!

**Review by snitin315 — APPROVED** [review]
LGTM, thank you for contributing ⭐ 

I'll leave it open for others to review.

File: lib/rules/no-constant-binary-expression.js [hunk 1]
```diff
@@ -45,6 +45,12 @@ function hasConstantNullishness(scope, node) {
             return (functionName === "Boolean" || functionName === "String" || functionName === "Number") &&
                 isReferenceToGlobalVariable(scope, node.callee);
         }
+        case "LogicalExpression": {
+            const isLeftConstant = hasConstantNullishness(scope, node.left);
+            const isRightConstant = hasConstantNullishness(scope, node.right);
+
+            return isLeftConstant || isRightConstant;
+        }
```
Comment (mdjermanovic) [inline lib/rules/no-constant-binary-expression.js:53]: These would be false positives, as the left-hand sides of these `??` do not have constant nullishness so the expressions neither always nor never short-circuit:

```js
const first = (foo && true) ?? bar;
//             ^^^^^^^^^^^ Unexpected constant nullishness on the left-hand side of a `??` expression.

const second = null ?? foo ?? bar;
//             ^^^^^^^^^^^ Unexpected constant nullishness on the left-hand side of a `??` expression.

const third = foo ?? null ?? bar;
//            ^^^^^^^^^^^ Unexpected constant nullishness on the left-hand side of a `??` expression.

```

We should check if the operator is `??` and if its right side is non-nullish.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: lib/rules/no-constant-binary-expression.js (hunk 1 above)
Comment (nissy-dev) [reply lib/rules/no-constant-binary-expression.js]: Thank you for review! I tried to fix in 7c91ba8.

File: docs/src/rules/no-constant-binary-expression.md [hunk 2]
```diff
@@ -53,6 +53,12 @@ const value4 = new Boolean(foo) === true;
 const objIsEmpty = someObj === {};
 
 const arrIsEmpty = someArr === [];
+
+const shortCircuit1 = condition1 && false && condition2;
+
+const shortCircuit2 = condition1 || true || condition2;
+
+const shortCircuit3 = condition1 && "non-nullish" && condition2;
```
Comment (mdjermanovic) [inline docs/src/rules/no-constant-binary-expression.md:61]:
```suggestion
const shortCircuit3 = condition1 ?? "non-nullish" ?? condition2;
```

I believe this was intended to be `??`. Otherwise, the example would be valid for this rule.

File: lib/rules/no-constant-binary-expression.js [hunk 3]
```diff
@@ -45,6 +62,9 @@ function hasConstantNullishness(scope, node) {
             return (functionName === "Boolean" || functionName === "String" || functionName === "Number") &&
                 isReferenceToGlobalVariable(scope, node.callee);
         }
+        case "LogicalExpression": {
+            return node.operator === "??" && (hasConstantNullishness(scope, node.right) && !isNullOrUndefined(scope, node.right));
+        }
```
Comment (mdjermanovic) [inline lib/rules/no-constant-binary-expression.js:67]: These would be false positives:

```js
/*eslint no-constant-binary-expression: "error"*/

const foo = a ?? (doSomething(), undefined) ?? b;
//          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

const bar = a ?? (something = null) ?? b;
//          ^^^^^^^^^^^^^^^^^^^^^^^

```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by nissy-dev** [comment]
Thank you for your review. I update my codes based on your comments. I would appreciate if you could review again.

**Review by mdjermanovic — COMMENTED** [review]
Just one small suggestion, then LGTM.

File: tests/lib/rules/no-constant-binary-expression.js [hunk 4]
```diff
@@ -308,6 +312,10 @@ ruleTester.run("no-constant-binary-expression", rule, {
         { code: "x === /[a-z]/", errors: [{ messageId: "alwaysNew" }] },
 
         // It's not obvious what this does, but it compares the old value of `x` to the new object.
-        { code: "x === (x = {})", errors: [{ messageId: "alwaysNew" }] }
+        { code: "x === (x = {})", errors: [{ messageId: "alwaysNew" }] },
+
+        { code: "window.abc && false && anything", errors: [{ messageId: "constantShortCircuit" }] },
+        { code: "window.abc || true || anything", errors: [{ messageId: "constantShortCircuit" }] },
+        { code: "window.abc ?? 'non-nullish' ?? 'non-nullish'", errors: [{ messageId: "constantShortCircuit" }] }
```
[the remark this answers is not in the document]
Comment (nissy-dev) [reply tests/lib/rules/no-constant-binary-expression.js]: Thank you for review! I committed your suggestion.

**Review by nzakas — APPROVED** [review]
LGTM pending @mdjermanovic's review.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!
